[RDY] vim-patch:7.4.1898 and man.vim modifiers support

nhooyr opened the pull request:
#### vim-patch:7.4.1898

Problem:    User commands don't support modifiers.
Solution:   Add the `<mods>` item. (Yegappan Lakshmanan, closes vim/vim#829)

https://github.com/vim/vim/commit/63a60ded3fd584847a05dccf058026e682abad90

nhooyr committed:
vim-patch:7.4.1898

Problem:    User commands don't support modifiers.
Solution:   Add the <mods> item. (Yegappan Lakshmanan, closes vim/vim#829)

https://github.com/vim/vim/commit/63a60ded3fd584847a05dccf058026e682abad90

nhooyr commented:
Once this is merged, I'll update man.vim.

nhooyr commented:
Also, I removed support for the `browse` modifier because our `cmdmod_T` doesn't include it.

https://github.com/neovim/neovim/blob/master/src/nvim/ex_cmds_defs.h#L168

nhooyr commented:
Try

```vim
:tab Man printf
q
Man printf
```

Not sure why but line numbers reappear? This [line](https://github.com/nhooyr/neovim/blob/3871f31ab1d795b0b73d87d5c3686d8ac75869ed/runtime/autoload/man.vim#L65) is what opens the manpage. I was under the assumption that we didn't have to set the window local options every time if we used `setlocal`. This only seems to happen with the `tab` modifier.

jamessan commented:
> I removed support for the browse modifier because our cmdmod_T doesn't include it.

`:browse` still exists in at least 0.1.4.  Maybe it's handled differently than in Vim, though.

> I was under the assumption that we didn't have to set the window local options every time if we used setlocal.

If you close a window and open a new one, then that's new window needs to have its local options set.

nhooyr commented:
If you do
```
set number
:Man printf
q
:Man printf
```
Even though a new window is opened at the second `:Man printf` and man.vim does not set any options again, line numbers do not reappear. Why is it different for `:tab Man printf`?

nhooyr commented:
Looks like we ignore `:browse` https://github.com/neovim/neovim/blob/add41dca98b748a84a022c828bc47e0e47e58332/src/nvim/ex_docmd.c#L1317-L1318

nhooyr commented:
man.vim doc update for this is in #5171

jamessan commented:
It's been like that since the initial import into @tarruda's repository.

I think the `:browse` command (guarded by the FEAT_BROWSE define) and the `:browse` command modifier (guarded by the FEAT_BROWSE_CMD define) were thought to require the GUI, but the modifier doesn't and some uses of the command don't either.  For example, `:browse oldfiles` uses the more prompt to let you select the file to open.  `:browse e` will use NetRW to let you select a file.

jamessan commented:
Ugh, apparently `:oldfiles!` was added to handle the file selection, which is bad because that means the only way to select a file is to make Vim abandon a non-hidden buffer (due to overloading the use of `eap->forceit`).

Also, we still document that `:browse oldfiles` will let you select the file (and will prompt if that would cause a modified buffer to be abandoned), while `:browse oldfiles!` forces a modified buffer to be abandoned.

nhooyr commented:
Alright, I've added `:browse` support.

jamessan commented:
> Alright, I've added :browse support.

I'm about to submit a PR that handles that independently.

nhooyr commented on src/nvim/ex_docmd.c:
```diff
@@ -5310,6 +5333,81 @@ uc_check_code (
+		  char *name;
+	  } mod_entry_T;
+	  static mod_entry_T mod_entries[] = {
+		  {&cmdmod.browse, "browse"},
```
@jamessan Where is this `cmdmod` variable defined? I cannot find it anywhere. I can only find `static struct cmdmod` on line 2524 but that doesn't seem to be it.

nhooyr committed:
man.vim: support for command modifiers

Closes #5235

nhooyr commented:
Requires #5214 to be merged first.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -56,20 +56,19 @@ function! man#open_page(count, count1, ...) abort
-  let found_man = s:find_man()
-  if getbufvar(bufname, 'manwidth') ==# s:manwidth()
-    if found_man
+  if s:find_man()
```
If `<mods>` is provided, does it make sense to still attempt to find a man window first and use that instead?

justinmk replied:
I think it should still try to find the closest window.

nhooyr replied:
Let me see.

nhooyr replied:
Not much code an done.

nhooyr replied:
I changed it to work just like `:help`. `:help` does not look for the closest window only if `tab` is in the modifiers.

nhooyr committed:
man.vim: set window local options when reusing buffer

This is necessary incase the buffer was previously opened in a different
tab, in which the window options there do not carry over. It is not
explicitly documented in ':help local-options' but that is how it works.

nhooyr commented:
This is ready.

nhooyr commented:
Filed vim/vim#976 for no mention of how tabs relate in `:help local-options`.

justinmk commented:
Build is failing. Does it need a rebase?

nhooyr commented:
Looks like something is wrong with the test.

jamessan commented:
The [lint] checks also fail.  It looks like you copied the code directly from Vim, instead of adapting it to Neovim's [coding style].

[lint]: https://travis-ci.org/neovim/neovim/jobs/151762844#L554
[coding style]: https://neovim.io/develop/style-guide.xml

nhooyr commented:
Fixed.

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -5311,6 +5333,86 @@ uc_check_code (
+    }
+
+    // TODO(nhooyr): How to support :noautocmd?
+    // TODO(nhooyr): How to support :sandbox
```
don't take ownership of upstream TODOs just to satisfy linter. We can just ignore those lint errors.

nhooyr replied:
Fixed.

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -5311,6 +5333,86 @@ uc_check_code (
+      { (bool *)&cmdmod.hide, "hide" },
+      { (bool *)&cmdmod.keepalt, "keepalt" },
+      { (bool *)&cmdmod.keepmarks, "keepmarks" },
+      { (bool *)&cmdmod.keepjumps, "keepjumps" },
```
keepjumps is above keepmarks in the original patch. why was it changed?

nhooyr replied:
fixed.

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -5311,6 +5333,86 @@ uc_check_code (
+  {
+    bool multi_mods = false;
+    typedef struct {
+      bool *varp;
```
changing this to bool just makes a mess, it's not gaining us anything.

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -5311,6 +5333,86 @@ uc_check_code (
+    }
+
+    // the modifiers that are simple flags
+    for (uint64_t i = 0; i < ARRAY_SIZE(mod_entries); i++) {
```
why `uint64_t` instead of `size_t`?

justinmk replied:
the lack of a `{NULL, NULL}` entry in `mod_entries` could introduce a bug when someone merges a future patch that uses the `mod_entries[i].varp != NULL` idiom.

nhooyr replied:
fixed.

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -5311,6 +5333,86 @@ uc_check_code (
+
+    // the modifiers that are simple flags
+    for (uint64_t i = 0; i < ARRAY_SIZE(mod_entries); i++) {
+      if (*mod_entries[i].varp) {
```
don't need to check this since you are checking `i < ARRAY_SIZE(mod_entries)` and we don't have the `{NULL, NULL}` entry.

nhooyr replied:
That checks if the modifier was set. Without it, we would be adding all of the modifiers in `mod_entries` every single time.

justinmk commented on src/nvim/testdir/test_usercommands.vim:
```diff
@@ -0,0 +1,44 @@
+  call assert_equal('"vertical" ', mods)
+
+  delcommand MyCmd
+  delcommand MyQCmd
```
why remove the `unlet g:mods` call? Even if it's not needed for some reason, avoid arbitrary differences, it makes merging and cross-reference painful.

justinmk commented on src/nvim/testdir/test_usercommands.vim:
```diff
@@ -0,0 +1,44 @@
+        \ 'belowright botright browse confirm hide keepalt keepmarks ' .
+        \ 'keepjumps keeppatterns lockmarks noswapfile silent tab topleft ' .
+        \ 'verbose vertical ', mods)
+  let mods = ''
```
original patch calls this `g:mods`, and that's part of our style guidelines anyways.

justinmk commented on src/nvim/testdir/test_usercommands.vim:
```diff
@@ -0,0 +1,44 @@
+        \ 'verbose vertical ', mods)
+  let mods = ''
+
+  command! -nargs=* MyQCmd let mods .= '<q-mods> '
```
`g:mods`

justinmk commented on src/nvim/testdir/test_usercommands.vim:
```diff
@@ -0,0 +1,44 @@
+function Test_cmdmods()
+  let mods = ''
+
+  command! -nargs=* MyCmd let mods .= '<mods> '
```
a command cannot modify a local variable. I actually am not sure how this test would pass.

oni-link commented on src/nvim/ex_docmd.c:
```diff
@@ -5311,6 +5333,87 @@ uc_check_code (
+    } mod_entry_T;
+    static mod_entry_T mod_entries[] = {
+      { &cmdmod.browse, "browse" },
+      { (bool *)&cmdmod.confirm, "confirm" },
```
Is the value pointed to not dependent on endianness and we could read the wrong value when dereferencing?

nhooyr replied:
I don't think so. The test passes fine so the correct value is read.

nhooyr replied:
Actually, maybe I should change the set member to have a type of `int *` then?

oni-link replied:
`noswapfile` and `browse` have type `bool` in `cmdmod_T`. Using an `int`-pointer to read them could also be problematic.

oni-link replied:
Why do we use a pointer instead of an `int` or `bool` for this member? Do we ever write to this member?

nhooyr replied:
How about I change every member of `cmdmod_T` to be a boolean? I'll go over the rest of the code to make sure it's fine.

nhooyr replied:
Looks like only the tab member is used as an actual integer.

nhooyr replied:
It's because its a static array. This way, we don't create a new array every single time the method is invoked. We just keep pointers to the fields of the global `cmdmod`.

nhooyr replied:
Oh tab is already special cased. I can just make them all booleans then.

oni-link commented on src/nvim/ex_docmd.c:
```diff
@@ -5154,6 +5154,25 @@ static char_u *uc_split_args(char_u *arg, size_t *lenp)
 }
 
+  static size_t
+add_cmd_modifier(char_u *buf, char *mod_str, bool *multi_mods)
```
Return type should not be on its own line.

nhooyr committed:
use bool type for flag mod_entry_T members

oni-link commented on src/nvim/ex_cmds_defs.h:
```diff
@@ -164,17 +164,17 @@ typedef struct expand {
+  bool keepjumps;    ///< TRUE when ":keepjumps" was used
+  bool keepmarks;    ///< TRUE when ":keepmarks" was used
+  bool keeppatterns; ///< TRUE when ":keeppatterns" was used
+  bool lockmarks;    ///< TRUE when ":lockmarks" was used
```
We use `true` instead of `TRUE` when the type of the variable is `bool`. Also in comments.

nhooyr replied:
Yup, I realized that in bed. Just got up to fix it :D

nhooyr committed:
man.vim: slight refactoring

Addresses problem one in #5240

nhooyr commented:
This regressed #5172. Fixing.

nhooyr committed:
man.vim: if reusing a buffer, do not use noautocmd

The commit that added support for modifiers regressed #5168
causing #5172. This commit fixes it again.

justinmk commented on runtime/autoload/man.vim:
```diff
@@ -49,48 +46,54 @@ function! man#open_page(count, count1, ...) abort
+      return
     endif
-    keepjumps 1
+    noautocmd execute 'edit' bufname
```
Why is `noautocmd` used? It should only be used if there was already a chance given to users/plugins to react to autocmds. Will user `FileType` autocmds be respected, for example?

nhooyr replied:
It is used so that the extraction and verification isn't repeated in `man#read_page` when the BufReadCmd autocmd is triggered. Yes, they will be respected because the filetype is set later in `s:read_page`.

nhooyr replied:
I just compressed this section a bit in the amended commit I pushed.

justinmk commented on src/nvim/ex_cmds_defs.h:
```diff
@@ -164,18 +164,18 @@ typedef struct expand {
+  bool keeppatterns;  ///< true when ":keeppatterns" was used
+  bool lockmarks;     ///< true when ":lockmarks" was used
+  bool noswapfile;    ///< true when ":noswapfile" was used
+  char_u *save_ei;    ///< saved value of 'eventignore'
```
We generally don't re-order things compare to the Vim source if we did not make any functional changes. It makes merging and comparing the Vim source difficult.

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -5311,6 +5332,87 @@ uc_check_code (
+        *buf++ = '"';
+      }
+      *buf = '\0';
+    }
```
Here again things were reordered but it's not clear why.

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -5311,6 +5332,87 @@ uc_check_code (
+    }
+
+    typedef struct {
+      bool *set;
```
Also wondering about this rename.

Basically, try to thing about whether it is worth the cost of having to cross-reference different names. If there is no major gain, it's better to just use the original name, if it's not idea. If you rewrote the function and cross-referencing is irrelevant, _then_ it might make sense to rename things.

justinmk commented on runtime/autoload/man.vim:
```diff
@@ -49,48 +46,52 @@ function! man#open_page(count, count1, ...) abort
+  " Ensure Vim is not recursively invoked (man-db does this)
+  " by forcing man to use cat as the pager.
+  " More info here http://comments.gmane.org/gmane.editors.vim.devel/29085
+  silent execute 'read!env MANPAGER=cat MANWIDTH='.b:manwidth s:man_cmd a:path
```
It's good practice to have a space after `:read`:

    read !env

because `:write` behaves differently based on that. `:read` doesn't, but... it's a good practice.

justinmk commented:
Verified test. Thanks @nhooyr !
